L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -86,53 +123,310 @@ HGCalConcentratorSuperTriggerCellImpl::getSuperTriggerCellId(int detid) const {
+                newtc = newtc | (k<<2);
+                
+                if ( !triggerTools_.validTriggerCell(newtc) ) {
+                  s.second.reject();
+                  continue;
+                }
+                
+                
+                l1t::HGCalTriggerCell triggerCell;
+                GlobalPoint point = triggerTools_.getTCPosition(newtc);
+                math::PtEtaPhiMLorentzVector p4(0, point.eta(), point.phi(), 0.);
+                
+                triggerCell.setPosition(point);     
+                triggerCell.setP4(p4);
+                triggerCell.setDetId(newtc);
+
+                s.second.addToList( triggerCell );                      
+                trigCellVecOutput.push_back ( triggerCell );
+                
+              }
+              
+            }
+            
+          }
+          
+        }
+        
+      }
+     
+
+ 
+    }
+
+        
+ 
+}
+
+
+void 
+HGCalConcentratorSuperTriggerCellImpl::
+coarsenTriggerCells( std::unordered_map<unsigned,SuperTriggerCell>& STCs, const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput)
+{ 
+  std::unordered_map<unsigned,SuperTriggerCell> coarseTCs;
+
+
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+    if (tc.subdetId() != HGCHEB) {
+      coarseTCs[getSuperTriggerCellId(tc.detId(),2)].add(tc);
+    }
+    else{
+      trigCellVecOutput.push_back( tc );
+    }
+  }
+
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+     if (tc.subdetId() == HGCHEB) continue;
+
+     int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+     if ( thickness > 0 ){
+
+      if (!(tc.detId() & 1)) { //if even
+        const auto & ctc = coarseTCs[getSuperTriggerCellId(tc.detId(),2)]; 
+        auto & stc = STCs[getSuperTriggerCellId(tc.detId())]; 
+        trigCellVecOutput.push_back( tc );
+        ctc.assignEnergy(trigCellVecOutput.back(), "STC");      
+        //        //reassign max id to the correct one;
+        if ( trigCellVecOutput.back().hwPt() >= stc.GetMaxHwPt() ){
+
+          if ( stc.GetMaxId() != tc.detId() ){
+            stc.SetMaxTC( trigCellVecOutput.back() );
+          }
+
+        }
+      }
+
+    }
+    else{
+
+      trigCellVecOutput.push_back( tc );
+
+    }
+
+
+  }
+
+
+}
+
+
 void 
 HGCalConcentratorSuperTriggerCellImpl::
 superTriggerCellSelectImpl(const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput)
 { 
 
+  fixedDataSize_ = true;
+
   std::unordered_map<unsigned,SuperTriggerCell> STCs;
+  std::vector<l1t::HGCalTriggerCell> trigCellVecInputEnlarged;
 
-  // first pass, fill the super trigger cells
+  // first pass, fill the "coarse" trigger cells
   for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+    trigCellVecInputEnlarged.push_back ( tc );
     if (tc.subdetId() == HGCHEB) continue;
     STCs[getSuperTriggerCellId(tc.detId())].add(tc);
```

Comment:
If now TCs are created from STCs, I don't understand why the input TCs need to be stored in the STC object

Reply:
The only reason would be to easily access other quantities such as thickness. I think it's a similar number of lines of code to recreate a TCid from a STC, or store the original number.

Reply:
In the near future, the TC<->STC mapping will have to be done somewhere else (e.g. in the geometry). In that case we need to compute these numbers on the fly, and not store them.
So to facilitate the future migration, it's better to compute STC id from TC id and vice-versa.

Reply:
Ok, but somehow we would need to store all the other information encoded in a TC, also in the STC (currently just the cell info).
Probably in the `getSuperTriggerCellId` function instead of
```
TC_split = (TC_idV8.cell() & kSplit_.at( stcSize_.at(thickness) ) );
return TC_wafer<<kWafer_offset_ | TC_split;
```
we would want
```
TC_split = INT_MAX & kSplit_.at( stcSize_.at(thickness) ) );
TC_split = (TC_idV8.cell() & TC_split;
return TC_split;
```
Is that correct?

Reply:
I don't think this is correct.
The best would be to use an `HGCalDetId`, with `TC_split` in place of the cell id. The id should be the same as the trigger cells inside STC except for the cell id

Reply:
I guess I'm not quite getting it - at this point we don't have access to the TC id (if we don't form the list), so can't form a HGCalDetId. We only have the TC_split number.

Reply:
The SuperTC class should store a complete id (HGCalDetId for V8) to be able to use it to compute the TC detids. i.e. the subdet, zside, layer, wafer, etc. information of the STC should be stored.